### PR TITLE
feat: (Core) component for object marker with document changes

### PR DESCRIPTION
**Pull request by Lokanathan-k**
#### Please provide a link to the associated issue.

https://github.com/SAP/fundamental-ngx/issues/2912

#### Please provide a brief summary of this pull request.

This indicates the technical status of an object.
It is enabled for the dynamic page layout, snapping header, object page header, upload collection, and object list item.
The technical status can be represented as and icon, with an icon and text, or as text only, depending on the screen size.
fd-object-marker is created as directive in the core library.
Currently, the following technical statuses can be visualized by the object marker:

> Flag
> Favorite
> Editing Status (Draft, Locked, Unsaved Changes)
> The editing status is part of the draft handling concept.

Screenshot 2020-09-01 at 11 07 41 AM

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-Icon-text-example.component.html
Comment (kavya-b) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-Icon-text-example.component.html:1]: `[glyph]="'add-favorite'"` can be `glyph="add-favorite"` right? I think for string values we are not keeping the binding.

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html
Comment (kavya-b) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html:1]: change [componentName]="'objectstatus'" to correct component

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-header/object-marker-header.component.html
Comment (kavya-b) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-header/object-marker-header.component.html:7]: unnecessary capitalization of every word

**Comment by kavya-b** [comment]
- Please elaborate on the examples descriptions. All are the same now.
</img>
- Is focus on each marker not needed? I see it on the linked marker in UI5
</img>
- when only glyphs are present, how is a11y supported? Shouldn't there be a tooltip on hover?  I see it present in UI5 samples.

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-Icon-text-example.component.html
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-Icon-text-example.component.html:1]: unnecessary bindings

Comment (Betrozov) [reply apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-Icon-text-example.component.html:1]: plz check it in the whole file

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-clickable-example.component.html
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-clickable-example.component.html:1]: the same

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.html
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.html:1]: the same

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html:1]: unnecessary bindings
plz check it in the whole file

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.ts
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.ts:13]: please use camelCase format
plz check it in the whole file

File: libs/core/src/lib/object-marker/object-marker.component.spec.ts
Comment (Betrozov) [inline libs/core/src/lib/object-marker/object-marker.component.spec.ts:4]: plz sort imports (angular → 3rd party or aliases → relative)

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.module.ts
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.module.ts:15]: plz sort imports (angular → 3rd party or aliases → relative)

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Betrozov) [inline libs/core/src/lib/object-marker/object-marker.component.ts:35]: `private` -> `private readonly`

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Betrozov) [inline libs/core/src/lib/object-marker/object-marker.component.ts:42]: should by `private`

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Betrozov) [inline libs/core/src/lib/object-marker/object-marker.component.ts:62]: seems it is not used

**Review by Betrozov — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Lokanathan-k) [reply libs/core/src/lib/object-marker/object-marker.component.ts:62]: Used for unit testing

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Lokanathan-k) [reply libs/core/src/lib/object-marker/object-marker.component.ts:42]: Signature of the method from class CssClassBuilder does not match

**Comment by Betrozov** [comment]
something broke icon styles
Screenshot 2020-09-16 at 14 17 25

**Comment by Lokanathan-k** [comment]
> something broke icon styles
> Screenshot 2020-09-16 at 14 17 25

@Betrozov Its the breaking change from the button.
https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes

**Comment by KevinOkamoto** [comment]
> @Betrozov Its the breaking change from the button.
> https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes

I think we need to address these breaking changes from fundamental-styles. We should either a) fix the breaking changes in this PR, or b) open another PR where we do the upgrade of fundamental-styles and fix the breaking changes.

Either way I don't think we should merge this PR until the breaking changes are addressed. I don't want to leave the master branch in a broken state.

**Comment by KevinOkamoto** [comment]
I had a conversation with @InnaAtanasova and @droshev about the fundamental-styles upgrade. They told me that the upgrade to 0.12 and fixes to the breaking changes will be addressed soon in a separate PR.

So we'll be able to merge this PR once the upgrade PR is merged.

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (KevinOkamoto) [inline libs/core/src/lib/object-marker/object-marker.component.ts:61]: Do we need to invoke `buildComponentCssClass()` for every change? And if so do we also need to update the glyph class on the icon that is created?

**Review by KevinOkamoto — COMMENTED** [review]
See comment.

**Review by Betrozov — APPROVED** [review]

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.ts
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.ts:27]: please use camelCase

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.ts
Comment (Betrozov) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.ts:26]: plz remove public it is unnecessary

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Betrozov) [inline libs/core/src/lib/object-marker/object-marker.component.ts:41]: plz add readonly for renderer

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (Betrozov) [inline libs/core/src/lib/object-marker/object-marker.component.ts:83]: plz change any to HTMLElement

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Review by Betrozov — APPROVED** [review]

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html
Comment (JKMarkowski) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html:45]: Could you add typescript file to example files? Now it doesn't support stackblitz

File: apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.ts
Comment (JKMarkowski) [inline apps/docs/src/app/core/component-docs/object-marker/examples/object-marker-example.component.ts:30]: In case you have some functionality, to keep stackblitz support you would need to put it to another file.

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html
Comment (JKMarkowski) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html:39]: `hoverig` => `hovering`

File: apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html
Comment (JKMarkowski) [inline apps/docs/src/app/core/component-docs/object-marker/object-marker-docs.component.html:17]: `hoverig` => `hovering`

File: libs/core/src/lib/object-marker/object-marker.component.scss
Comment (JKMarkowski) [inline libs/core/src/lib/object-marker/object-marker.component.scss:1]: import of Icons is not needed there, because it's imported already in `angular.json` and will be imported that way by end-users. It's just to reduce size of output file.

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (JKMarkowski) [inline libs/core/src/lib/object-marker/object-marker.component.ts:55]: You can put `<i class="fd-object-marker__icon sap-icon--" *ngIf="glyph">` into template. This approach is quite ugly IMO

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (JKMarkowski) [inline libs/core/src/lib/object-marker/object-marker.component.ts:67]: With my approach of having icons in template, you would be able to get rid of this

File: package.json
Comment (JKMarkowski) [inline package.json:102]: Let's revert this change and include css code from `object-marker.css` into `object-marker.component.scss`

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
Good Job! Let me know if something is not clear

File: package.json
Comment (Lokanathan-k) [reply package.json:102]: This PR will merged post the Breaking changes are fixed if i revert this change build will fail. this is already conveyed to @InnaAtanasova

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
There is supposed to be some padding between the icon and the text.  Also the links should be underlined when hovered. https://fundamental-styles.netlify.app/?path=/docs/components-object-marker--icon-only#clickable-object-marker

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (JKMarkowski) [inline libs/core/src/lib/object-marker/object-marker.component.ts:11]: Could you put `Renderer2` import to same import as others?

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
image

The `fd-object-marker__text` element still exist, even if there is no text. There is no way to make it work with content projection. Can you use solution, which has been introduced there https://github.com/SAP/fundamental-ngx/pull/3434 ?

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (JKMarkowski) [inline libs/core/src/lib/object-marker/object-marker.component.ts:41]: Is there any reason that those variables are there?

File: apps/docs/src/app/core/api-files.ts
Comment (JKMarkowski) [inline apps/docs/src/app/core/api-files.ts:163]: Let's remove module from this array

**Review by JKMarkowski — APPROVED** [review]
Could you also remove `package-lock.json` changes? I don't see any changes in `package.json`, so it's not necessary
Besides this and 1 comment below, looks good to me. Good Job :)

File: libs/core/src/lib/object-marker/object-marker.component.ts
Comment (JKMarkowski) [inline libs/core/src/lib/object-marker/object-marker.component.ts:30]: Just one thing regarding comments. Could you also add 1 line of space between them and change:
```    
/**
     * Glyph (icon) of the Object Status.
 */
```
To:
`/** Glyph (icon) of the Object Marker */`

**Review by JKMarkowski — CHANGES_REQUESTED** [review]
Let me know if the changes are applied, so I will approve

**Review by KevinOkamoto — APPROVED** [review]
Looks good!

**Review by JKMarkowski — APPROVED** [review]

**Review by mikerodonnell89 — APPROVED** [review]

**Pull request merged**
